prov/gni/src/gnix_fabric.c

```diff
@@ -660,11 +692,187 @@ __gnix_fab_ops_set_val(struct fid *fid, fab_ops_val_t t, void *val)
+	if (!info)
+		return -FI_ENOMEM;
+
+	if (attr)
+		info->attr = *attr;
+	else {	
+		info->attr.user_key_limit = gnix_default_user_registration_limit;
+		info->attr.prov_key_limit = gnix_default_prov_registration_limit;
+	}	
+
+	ret = _gnix_auth_key_insert(auth_key, auth_keylen, info);
+	if (ret) {
+		GNIX_INFO(FI_LOG_FABRIC, "failed to insert authorization key"
+			", key=%p len=%d ret=%d\n", 
+			auth_key, auth_keylen, ret);
+		_gnix_auth_key_free(info);
+		info = NULL;
+	}
+
+	return ret;
+}
+
+static int __gnix_auth_key_set_val(uint8_t *auth_key, size_t auth_keylen, gnix_auth_key_opt_t opt, void *val)
+{
+	struct gnix_auth_key *info;
+	int v;	
+	int ret;
+	
+	if (!val)
+		return -FI_EINVAL;
+
+	if (auth_keylen == GNIX_PROV_DEFAULT_AUTH_KEYLEN) {
+		info = _gnix_auth_key_lookup(auth_key, auth_keylen);
+	} else {
+		GNIX_INFO(FI_LOG_FABRIC, "multiple authorization keys "
+			"not supported yet\n");
+		return -FI_ENOSYS; /* no support for out-of-band auth keys yet */
+	}
+
+	if (!info) {
+		ret = __gnix_auth_key_initialize(auth_key, auth_keylen, NULL);
+		assert(ret == FI_SUCCESS);
+
+		info = _gnix_auth_key_lookup(auth_key, auth_keylen);
+		assert(info);
+	}
+
+	/* if the limits have already been set, and the user is trying to modify it, kick it back */
+	if (opt == GNIX_USER_KEY_LIMIT || opt == GNIX_PROV_KEY_LIMIT) {
+		fastlock_acquire(&info->lock);
+		if (info->enabled) {
+			fastlock_release(&info->lock);
+			GNIX_INFO(FI_LOG_FABRIC, "authorization key already "
+				"enabled and cannot be modified\n");
+			return -FI_EAGAIN;
+		}
+	}
+
+	switch (opt) {
+	case GNIX_USER_KEY_LIMIT:
+		v = *(int *) val;
+		if (v >= GNIX_MAX_SCALABLE_REGISTRATIONS) {
+			GNIX_ERR(FI_LOG_FABRIC, "User is requesting more registrations than is present on node\n");
+			ret = -FI_EINVAL;
+		} else
+			info->attr.user_key_limit = v;
+		fastlock_release(&info->lock);
+		break;
+	case GNIX_PROV_KEY_LIMIT:
+		v = *(int *) val;
+		if (v >= GNIX_MAX_SCALABLE_REGISTRATIONS) {
+			GNIX_ERR(FI_LOG_FABRIC, "User is requesting more registrations than is present on node\n");
+			ret = -FI_EINVAL;
+		}
+		info->attr.prov_key_limit = v;
+		fastlock_release(&info->lock);
+		break;
+	case GNIX_TOTAL_KEYS_NEEDED:
+		GNIX_WARN(FI_LOG_FABRIC, "GNIX_TOTAL_KEYS_NEEDED is not a definable value.\n");
+		return -FI_EINVAL;
 	default:
 		GNIX_WARN(FI_LOG_FABRIC, ("Invalid fab_ops_val\n"));
 		return -FI_EINVAL;
 	}
 
+	return ret;
+}
+
+static int __gnix_auth_key_get_val(uint8_t *auth_key, size_t auth_keylen, gnix_auth_key_opt_t opt, void *val)
+{
+	struct gnix_auth_key *info;
+	int ret;
+
+	if (!val)
+		return -FI_EINVAL;
+
+	if (auth_keylen == GNIX_PROV_DEFAULT_AUTH_KEYLEN) {
+		info = _gnix_auth_key_lookup(auth_key, auth_keylen);
+	} else {
+		GNIX_INFO(FI_LOG_FABRIC, "multiple authorization keys not yet"
+			"supported\n");
+		return -FI_ENOSYS; 
+	}
+
+	if (!info) {
+		ret = __gnix_auth_key_initialize(auth_key, auth_keylen, NULL);
+		assert(ret == FI_SUCCESS);
```

Comment:
My understanding is that these asserts are compiled out without debugging enabled. Maybe use `GNIX_FATAL` instead?